File: practica Gral/polimorfismo/Mp4.py
#clase base
class Mp4:
    def __init__(self, marca, capacidad_gb, max_canciones, max_videos):
        self.marca = marca
        self.capacidad_gb = capacidad_gb
        self.nro_canciones = 0
        self.nro_videos = 0
        self.max_canciones = max_canciones
        self.max_videos = max_videos
        self.canciones = []
        self.videos = []

    def borrar_cancion_por_nombre(self, nombre):
        for i, c in enumerate(self.canciones):
            if c[0].lower() == nombre.lower():
                self.canciones.pop(i)
                self.nro_canciones -= 1
                print(f"Canción {nombre} eliminada por nombre.")
                return

    def borrar_cancion_por_artista(self, artista):
        for i, c in enumerate(self.canciones):
            if c[1].lower() == artista.lower():
                self.canciones.pop(i)
                self.nro_canciones -= 1
                print(f"Canción eliminada del artista: {artista}")
                return

    def borrar_cancion_por_peso(self, peso):
        for i, c in enumerate(self.canciones):
            if c[2] == peso:
                self.canciones.pop(i)
                self.nro_canciones -= 1
                print(f"Canción eliminada con peso: {peso}")
                return

    def add_cancion(self, nombre, artista, peso):
        if self.nro_canciones < self.max_canciones:
            self.canciones.append([nombre, artista, peso])
            self.nro_canciones += 1
            print(f"Canción añadida: {nombre}")
        else:
            print("No hay espacio para más canciones.")

    def add_video(self, nombre, artista, peso):
        if self.nro_videos < self.max_videos:
            self.videos.append([nombre, artista, peso])
            self.nro_videos += 1
            print(f"Video añadido: {nombre}")
        else:
            print("No hay espacio para más videos.")

    def capacidad_disponible(self):
        total_kb = sum(float(c[2]) for c in self.canciones)
        total_mb = sum(float(v[2]) for v in self.videos)

        total_gb = (total_kb / (1024 * 1024)) + (total_mb / 1024)
        disponible = self.capacidad_gb - total_gb

        print(f"Capacidad total: {self.capacidad_gb} GB")
        print(f"Usado: {total_gb:.3f} GB")
        print(f"Disponible: {disponible:.3f} GB")

    def mostrar_datos(self):
        print(f"Marca: {self.marca} | Capacidad: {self.capacidad_gb} GB")
        print("\nCanciones:")
        for c in self.canciones:
            print(f"- {c[0]} | {c[1]} | {c[2]} KB")
        print("\nVideos:")
        for v in self.videos:
            print(f"- {v[0]} | {v[1]} | {v[2]} MB")


# --- MAIN ---
if __name__ == "__main__":
    mp4 = Mp4("Sony", 8, 100, 100)

    mp4.add_cancion("Back To Black", "Amy Winehouse", "100")
    mp4.add_cancion("Lost On You", "LP", "150")

    mp4.add_video("Heathens", "Twenty One Pilots", "50")
    mp4.add_video("Without Me", "Halsey", "30")

    mp4.mostrar_datos()
    mp4.borrar_cancion_por_nombre("Lost On You")

    mp4.capacidad_disponible()

File: practica Gral/polimorfismo/PoliEj13/Mp4.java
package PoliEj13;
public class Mp4 {
    private String marca;
    private double capacidadGb;
    private int nroCanciones;
    private int nroVideos;
    private String[][] canciones;
    private String[][] videos;

    public Mp4(String marca, double capacidadGb, int maxCanciones, int maxVideos) {
        this.marca = marca;
        this.capacidadGb = capacidadGb;
        this.nroCanciones = 0;
        this.nroVideos = 0;
        this.canciones = new String[maxCanciones][3];
        this.videos = new String[maxVideos][3];
    }

    public void borrarCancionPorNombre(String nombre) {
        for (int i = 0; i < nroCanciones; i++) {
            if (canciones[i][0].equalsIgnoreCase(nombre)) {
                canciones[i] = canciones[nroCanciones - 1];
                canciones[nroCanciones - 1] = null;
                nroCanciones--;
                System.out.println("Canción " + nombre + " eliminada por nombre.");
                return;
            }
        }
    }

    public void borrarCancionPorArtista(String artista) {
        for (int i = 0; i < nroCanciones; i++) {
            if (canciones[i][1].equalsIgnoreCase(artista)) {
                canciones[i] = canciones[nroCanciones - 1];
                canciones[nroCanciones - 1] = null;
                nroCanciones--;
                System.out.println("Canción eliminada del artista: " + artista);
                return;
            }
        }
    }

    public void borrarCancionPorPeso(String peso) {
        for (int i = 0; i < nroCanciones; i++) {
            if (canciones[i][2].equalsIgnoreCase(peso)) {
                canciones[i] = canciones[nroCanciones - 1];
                canciones[nroCanciones - 1] = null;
                nroCanciones--;
                System.out.println("Canción eliminada con peso: " + peso);
                return;
            }
        }
    }

    public void addCancion(String nombre, String artista, String peso) {
        if (nroCanciones < canciones.length) {
            canciones[nroCanciones][0] = nombre;
            canciones[nroCanciones][1] = artista;
            canciones[nroCanciones][2] = peso;
            nroCanciones++;
            System.out.println("Canción añadida: " + nombre);
        } else {
            System.out.println("No hay espacio para más canciones.");
        }
    }

    public void addVideo(String nombre, String artista, String peso) {
        if (nroVideos < videos.length) {
            videos[nroVideos][0] = nombre;
            videos[nroVideos][1] = artista;
            videos[nroVideos][2] = peso;
            nroVideos++;
            System.out.println("Video añadido: " + nombre);
        } else {
            System.out.println("No hay espacio para más videos.");
        }
    }

    public void capacidadDisponible() {
        double totalKB = 0;
        double totalMB = 0;

        for (int i = 0; i < nroCanciones; i++) {
            totalKB += Double.parseDouble(canciones[i][2]);
        }

        for (int i = 0; i < nroVideos; i++) {
            totalMB += Double.parseDouble(videos[i][2]);
        }

        double totalGB = (totalKB / (1024 * 1024)) + (totalMB / 1024);
        double disponible = capacidadGb - totalGB;

        System.out.println("Capacidad total: " + capacidadGb + " GB");
        System.out.println("Usado: " + totalGB + " GB");
        System.out.println("Disponible: " + disponible + " GB");
    }

    public void mostrarDatos() {
        System.out.println("Marca: " + marca + " | Capacidad: " + capacidadGb + "GB");
        System.out.println("\nCanciones:");
        for (int i = 0; i < nroCanciones; i++) {
            System.out.println("- " + canciones[i][0] + " | " + canciones[i][1] + " | " + canciones[i][2] + "KB");
        }
        System.out.println("\nVideos:");
        for (int i = 0; i < nroVideos; i++) {
            System.out.println("- " + videos[i][0] + " | " + videos[i][1] + " | " + videos[i][2] + "MB");
        }
    }
}

